c++/cpp2py/converters/std_array.hpp

```diff
@@ -1,54 +1,118 @@
 #pragma once
 #include <array>
+#include <numpy/arrayobject.h>
+
+#include "../numpy_proxy.hpp"
+#include "../py_converter.hpp"
+#include "./vector.hpp"
 
 namespace cpp2py {
 
-  template <typename T, size_t R> struct py_converter<std::array<T, R>> {
-    // --------------------------------------
+  template <typename T, size_t R> numpy_proxy make_numpy_proxy_from_array(std::array<T, R> arr) {
 
-    static PyObject *c2py(std::array<T, R> const &v) {
-      PyObject *list = PyList_New(0);
-      for (auto const &x : v) {
-        pyref y = py_converter<T>::c2py(x);
-        if (y.is_null() or (PyList_Append(list, y) == -1)) {
-          Py_DECREF(list);
-          return NULL;
-        } // error
-      }
-      return list;
+    if constexpr (has_npy_type<T>) {
+      auto *arr_heap = new std::array<T, R>{std::move(arr)};
+      auto capsule   = PyCapsule_New(arr_heap, "guard", &delete_pycapsule<std::array<T, R>>);
+
+      return {1, // rank
+              npy_type<std::remove_const_t<T>>,
+              (void *)arr_heap->data(),
+              std::is_const_v<T>,
+              v_t{static_cast<long>(R)}, // extents
+              v_t{sizeof(T)},            // strides
+              capsule};
+    } else {
+      std::array<pyref, R> arr_obj(arr.size());
+      std::transform(begin(arr), end(arr), begin(arr_obj), [](auto &&x) { return py_converter<std::decay_t<T>>::c2py(std::move(x)); });
+      return make_numpy_proxy_from_array(std::move(arr_obj));
+    }
+  }
+
+  // Make a new array from numpy view
+  template <typename T, size_t R> std::array<T, R> make_array_from_numpy_proxy(numpy_proxy const &p) {
+    EXPECTS(p.extents.size() == 1);
+    EXPECTS(p.extents[0] == R);
+
+    std::array<T, R> arr{};
+
+    if (p.element_type == npy_type<pyref>) {
+      auto **data = static_cast<PyObject **>(p.data);
```

Comment:
I agree. In fact we have to be very careful not to ever add any additional members to the pyref.
This would break the conversion to a numpy array of objects.